src/llama.cpp

```diff
@@ -2173,18 +2202,53 @@ struct llama_hparams {
         return false;
     }
 
-    uint32_t n_gqa() const {
+    // TODO: deduplicate per-layer getters
+    uint32_t n_head_l(uint32_t layer) const {
```

Comment:
We can remove `hparams.n_head` and rename this to:

```c++
uint32_t n_head(uint32_t layer) const {
```

This way everywhere we reference `hparams.n_head`, we will use `hparams.n_head()`.
We can do this for the rest of the parameters that make sense to be per-layer, the main goal being to avoid having duplicated information in `hparams.n_head` and `hparams.n_head_vec`.

We can do this change in a follow-up refactoring PR, together with finding a more compact way to have per-layer hparams (maybe via new `struct llama_hparams_layer`)

Reply:
The advantage of duplicating the information is to have a fallback value, and it also avoids having to allocate the vectors when there's only one value for all the layers. The disadvantage is the possible confusion of which value to use.

> We can do this change in a follow-up refactoring PR, together with finding a more compact way to have per-layer hparams (maybe via new `struct llama_hparams_layer`)

Yes a follow-up refactor seems appropriate. The goal with this initial implementation was to minimize having to change the accesses to `hparams.n_head`, `hparams.n_head_kv`, and `hparams.n_ff` all over the place, and also to re-use the nice abstraction which `llama_model_loader.get_arr` seemed to be (it was introduced in #7225, but isn't used by anything on `master`).

Although a possible huge problem with using `std::vector` in `llama_hparams` is that it's no longer trivially copyable as assumed by `llama_state_save_file_internal` and `llama_state_load_file_internal`.

Reply:
> Although a possible huge problem with using std::vector in llama_hparams is that it's no longer trivially copyable as assumed by llama_state_save_file_internal and llama_state_load_file_internal.

Hm, correct. We could switch to `uint32 n_head_vec[LLAMA_MAX_LAYERS];` instead and use `n_layer` as the container size?